src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java

```diff
@@ -166,13 +246,43 @@ private Response createSession(Command command) throws IOException {
         setCommandCodec(new AppiumW3CHttpCommandCodec());
         refreshAdditionalCommands();
         setResponseCodec(dialect.getResponseCodec());
-        return result.createResponse();
+        Response response = result.createResponse();
+        if (this.appiumClientConfig != null && this.appiumClientConfig.isDirectConnectEnabled()) {
+            setDirectConnect(response);
+        }
+
+        return response;
     }
 
     public void refreshAdditionalCommands() {
         getAdditionalCommands().forEach(this::defineCommand);
     }
 
+    @SuppressWarnings("unchecked")
+    private void setDirectConnect(Response response) throws SessionNotCreatedException {
+        Map<String, ?> responseValue = (Map<String, ?>) response.getValue();
+
+        DirectConnect directConnect = new DirectConnect(responseValue);
+
+        if (!directConnect.isValid()) {
+            return;
+        }
+
+        if (!directConnect.protocol.equals("https")) {
+            throw new SessionNotCreatedException(
+                    String.format("The protocol must be https. %s was given.", directConnect.protocol));
```

Comment:
I think the error message could be improved, for example:

The '%s' protocol is the direct connection url %s returned by the remote server is not acceptable. The only supported protocol is 'https'